android/modules/ui/src/java/ti/modules/titanium/ui/widget/listview/TiListView.java

```diff
@@ -492,57 +485,45 @@ private void update(String query)
 			item.getProperties().put(TiC.PROPERTY_FOOTER_TITLE, properties.get(TiC.PROPERTY_FOOTER_TITLE));
 			item.getProperties().put(TiC.PROPERTY_FOOTER_VIEW, properties.get(TiC.PROPERTY_FOOTER_VIEW));
 
-			item.setParent(proxy);
+			item.setParent(this.proxy);
 			this.items.add(item);
 		}
 
-		// Pre-load views for smooth initial scroll.
-		final int preloadSize = Math.min(this.items.size(), PRELOAD_SIZE);
-		for (int i = 0; i < preloadSize; i++) {
-			this.items.get(i).getOrCreateView();
-		}
+		// Pre-load items of empty list.
+		if (shouldPreload) {
+			final int preloadSize = Math.min(this.items.size(), PRELOAD_SIZE);
 
-		// Update models.
-		updateModels();
-	}
+			for (int i = 0; i < preloadSize; i++) {
 
-	/**
-	 * Update adapter models.
-	 */
-	public void updateModels()
-	{
-		// Amend final index of items.
-		int i = 0;
-		for (ListItemProxy item : this.items) {
-			if (item.isPlaceholder()) {
-				continue;
+				// Pre-load views for smooth initial scroll.
+				this.items.get(i).getOrCreateView();
 			}
-
-			// Update item index, ignoring placeholder entries.
-			item.index = i++;
 		}
 
-		final Activity activity = TiApplication.getAppCurrentActivity();
-		final View previousFocus = activity != null ? activity.getCurrentFocus() : null;
-
-		// Notify the adapter of changes.
+		// Notify adapter of changes on UI thread.
 		this.adapter.notifyDataSetChanged();
 
-		// FIXME: This is not an ideal workaround for an issue where recycled items that were in focus
+		// FIXME: This is not an ideal workaround for an issue where recycled rows that were in focus
 		//        lose their focus when the data set changes. There are improvements to be made here.
 		//        This can be reproduced when setting a Ti.UI.TextField in the Ti.UI.ListView.headerView for search.
-		new Handler().post(new Runnable()
-		{
-			public void run()
+		final Activity activity = TiApplication.getAppCurrentActivity();
+		final View previousFocus = activity.getCurrentFocus();
+
+		if (activity != null && previousFocus != null) {
```

Comment:
We should check if `activity` is `null` before calling `activity.getCurrentFocus()` above.